pkg/ca/tinkca/signer.go

```diff
@@ -18,118 +18,66 @@ import (
-		}
-		// https://github.com/tink-crypto/tink-go/blob/0aadc94a816408c4bdf95885b3c9860ecfd55fc0/signature/subtle/ed25519_signer.go#L27
-		p := ed25519.NewKeyFromSeed(privKey.GetKeyValue())
-		return p, nil
+func curveFromTinkECDSACurveType(curveType tinkecdsa.CurveType) (elliptic.Curve, error) {
+	switch curveType {
+	case tinkecdsa.NistP256:
+		return elliptic.P256(), nil
+	case tinkecdsa.NistP384:
+		return elliptic.P384(), nil
+	case tinkecdsa.NistP521:
+		return elliptic.P521(), nil
 	default:
-		return nil, fmt.Errorf("unsupported key type: %s", k.GetTypeUrl())
+		// Should never happen.
+		return nil, fmt.Errorf("unsupported curve: %v", curveType)
 	}
 }
 
-// getPrimaryKey returns the first enabled key from a keyset.
-func getPrimaryKey(ks *tinkpb.Keyset) *tinkpb.KeyData {
-	for _, k := range ks.GetKey() {
-		if k.GetKeyId() == ks.GetPrimaryKeyId() && k.GetStatus() == tinkpb.KeyStatusType_ENABLED {
-			return k.GetKeyData()
-		}
+// KeyHandleToSigner constructs a [crypto.Signer] from a Tink [keyset.Handle]'s
+// primary key.
+//
+// NOTE: Tink validates keys on [keyset.Handle] creation.
+func KeyHandleToSigner(kh *keyset.Handle) (crypto.Signer, error) {
+	primary, err := kh.Primary()
+	if err != nil {
+		return nil, err
 	}
-	return nil
-}
 
-// validateEcdsaPrivKey validates the given ECDSAPrivateKey.
-// https://github.com/tink-crypto/tink-go/blob/0aadc94a816408c4bdf95885b3c9860ecfd55fc0/signature/ecdsa/signer_key_manager.go#L151
-func validateEcdsaPrivKey(key *ecdsapb.EcdsaPrivateKey) error {
-	if err := keyset.ValidateKeyVersion(key.Version, uint32(ecdsaSignerKeyVersion)); err != nil {
-		return fmt.Errorf("ecdsa: invalid key version in key: %s", err)
-	}
-	if err := keyset.ValidateKeyVersion(key.GetPublicKey().GetVersion(), uint32(ecdsaSignerKeyVersion)); err != nil {
-		return fmt.Errorf("ecdsa: invalid public version in key: %s", err)
-	}
-	hash, curve, encoding := getECDSAParamNames(key.PublicKey.Params)
-	return signatureSubtle.ValidateECDSAParams(hash, curve, encoding)
-}
+	switch privateKey := primary.Key().(type) {
+	case *tinkecdsa.PrivateKey:
+		publicKey, err := privateKey.PublicKey()
+		if err != nil {
+			return nil, err
+		}
+		ecdsaPublicKey := publicKey.(*tinkecdsa.PublicKey)
 
-// getECDSAParamNames returns the string representations of each parameter in
-// the given ECDSAParams.
-// https://github.com/tink-crypto/tink-go/blob/0aadc94a816408c4bdf95885b3c9860ecfd55fc0/signature/ecdsa/proto.go#L24
-func getECDSAParamNames(params *ecdsapb.EcdsaParams) (string, string, string) {
-	hashName := commonpb.HashType_name[int32(params.GetHashType())]
-	curveName := commonpb.EllipticCurveType_name[int32(params.GetCurve())]
-	encodingName := ecdsapb.EcdsaSignatureEncoding_name[int32(params.GetEncoding())]
-	return hashName, curveName, encodingName
-}
+		curve, err := curveFromTinkECDSACurveType(ecdsaPublicKey.Parameters().(*tinkecdsa.Parameters).CurveType())
```

Comment:
not sure if a defensive check for `ecdsaPublicKey.Parameters()` being `nil` is worth it here